### PR TITLE
Terraform Migration

### PR DESCRIPTION
**Pull request by tr0njavolta**

### CODE REVIEW

File: content/reference/crossplane-vs-upbound.md [hunk 1]
```diff
@@ -41,4 +41,3 @@ Upbound provides a cloud-native experience for Crossplane. Upbound supplements t
 ## Constraints
 
 Because Upbound fully manages control planes, Upbound locks down the hosted control plane environment to prevent malicious actors. You may not install arbitrary workloads (Pods, Deployments, etc) on your managed control plane.
-
```
Comment (ytsarev) [inline content/reference/crossplane-vs-upbound.md:44]: Is it an intentional new empty line?

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    	ami                   	= "ami-065deacbcaac64cf2"
+    	instance_type         	= "t2.micro"
+    	tags = {
+        	Name = var.vmName,
+    	}
+  	}
+	variable "vmName" {
+    	description = "VM name"
+    	type    	= string
+  	}
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:61]: the hcl is missing indentation

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+  forProvider:
+	source: Inline
+	module: |
+
+	###
+
+	vars:
+  	- key:
+    	value:
+```
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:90]: yaml is missing indentation, it's best to run yaml examples through https://github.com/adrienverge/yamllint

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+kind: Workspace
+metadata:
+  name: tf-vm
+spec:
+  forProvider:
+	source: Inline
+	module: |
+
+	resource "aws_instance" "my_vm" {
+    	ami                   	= "ami-065deacbcaac64cf2"
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:131]: is it some well-known AMI identifier? where is it coming from?

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+  	}
+	variable "vmName" {
+    	description = "VM name"
+    	type    	= string
+  	}
+
+	vars:
+  	- key: vmName
+    	value: crossplanevm
+```
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:145]: hcl indentation required

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+Copy and paste the configuration below.
+
+```yaml
+apiVersion: pkg.crossplane.io/v1
+kind: Provider
+metadata:
+  name: provider-terraform
+spec:
+  package: xpkg.upbound.io/upbound/provider-terraform
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:163]: it looks like the tag is missing, e.g.
```suggestion
  package: xpkg.upbound.io/upbound/provider-terraform:v0.13.0
```

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+`$ kubectl apply -f provider-terraform-install.yaml`
+
+Verify the provider with `kubectl get pods`.
+
+```yaml
+$ kubectl get pods -n upbound-system
+NAME                                                              READY   STATUS    RESTARTS   AGE
+crossplane-6979f579f9-x7nkr                                       2/2     Running   0          31m
+upbound-provider-aws-ec2-64262f355830-7b6b976c66-4b5wt            1/1     Running   0          29m
+upbound-provider-family-aws-fec919bd2218-5f85944578-rdxjm         1/1     Running   0          29m
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:180]: this list of pods does not look related to provider-terraform installation

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+```
+
+
+## Authenticate with your cloud provider
+
+The Crossplane AWS provider configuration handles authentication. You must
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:186]: The switch to aws provider is sudden here. Wasn't this story focused on provider-terraform?

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+create a Kubernetes secret file to authenticate with your AWS account.
+
+Crossplane supports AWS authentication with:
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:189]: To be precise specific provider supports it ( provider-aws or provider-terraform)

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+Authentication Keys
+Web Identity
+Service Accounts
+
+For more information on cloud provider authentication, checkout the Provider
+Azure and Provider GCP authentication documentation.
+
+In this example, the credentials.source attribute references the authentication
+key method as a secret.
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:198]: Is this copy-paste from some existing provider docs? looks alien here

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+## Deploy your configuration
+
+Next, apply the manifest.
+
+`$ kubectl apply -f terraform-config.yaml`
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:204]: it was named `terraform-configuration.yaml` earlier at https://github.com/upbound/docs/pull/290/files#diff-5b8369e7c3686919cb849742b6bee89edd3da0a788de313c8589ee70699f2f0bR117
```suggestion
`$ kubectl apply -f terraform-configuration.yaml`
```

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+      	secret_suffix 	= "providerconfig"
+      	namespace     	= "upbound-system"
+      	in_cluster_config = true
+    	}
+  	}
+  	provider "aws" {
+    	shared_credentials_files = ["${path.module}/aws-creds.ini"]
+    	region = "us-east-1"
+  	}
+```
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:246]: indentation required

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+`$ kubectl apply -f terraform-config.yaml`
+
+## State management
+
+For greenfield deployments, this approach works well for users who are already
+familiar with Terraform. If your organization wants to maintain state parity
+with a remote Terraform state file, you can use a `ProviderConfig` to point your
+Crossplane-controlled Terraform provider to the correct backend.
+
+The example below uses the `kubernetes` backend:
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:213]: `kubernetes` or other backend configuration is required before applying the Workspace MR, otherwise it fails. So maybe this part should be configured before applying Workspace in the previous step.

File: content/providers/provider-terraform/migrate-provider-tf.md [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+      	}
+    	}
+    	backend "kubernetes" {
+      	secret_suffix 	= "providerconfig"
+      	namespace     	= "upbound-system"
+      	in_cluster_config = true
+    	}
+  	}
+  	provider "aws" {
+    	shared_credentials_files = ["${path.module}/aws-creds.ini"]
```
Comment (ytsarev) [inline content/providers/provider-terraform/migrate-provider-tf.md:243]: explanation of this credentials file creation required

**Review by ytsarev — CHANGES_REQUESTED** [review]

File: content/providers/provider-terraform/migrate-hcl.md [hunk 15]
```diff
@@ -0,0 +1,527 @@
...
+<!-- vale Google.Will = NO -->
+
+<!-- vale gitlab.FutureTense = NO -->
+
+The `forProvider` sub-field defines the information you need for the instance
+configuration. Instead of relying on the Terraform configuration to define how
+you want to configure the instance, you'll use the Kubernetes manifest
+configuration language.
+
+## Create a resource definition
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:112]:
```suggestion
## Create a Crossplane custom resource definition
```

Should the headline be about the Crossplane object being created?

File: content/providers/provider-terraform/migrate-hcl.md [hunk 16]
```diff
@@ -0,0 +1,527 @@
...
+
+Your Composition sets the explicit values you want to deploy based on the
+definition file you created in the preceding step.
+
+## Create a claim
+
+Now that you have a definition and a composition, you can create a claim to. The
+claim is a set of resources to deploy within a single namespace. Creating claims
+is comparable to different Terraform workspaces in that resources within one
+namespace don't impact resources within another.
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:410]:
```suggestion
Now that you have a custom resource definition and a composition, you can create a claim and provision the resources. 
A claim deploys a set of resources within a namespace. Creating claims
is comparable to different Terraform workspaces. Resources in one
namespace don't impact resources in another namespace.
```

File: content/providers/provider-terraform/migrate-hcl.md [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+  parameters:
+    region: us-east-1
+    amiId: ami-0005e0cfe09cc9050
+    keyPairName: test-keypair
+    publicKey: |
+      ssh-rsa AAAAB3NzaC1...
+```
+
+Copy and paste the contents of your public key (typically `~/.ssh/id_rsa.pub`)
+into the `publicKey` field.
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:428]:
```suggestion
{{< hint "note" >}}
Copy and paste the contents of your public key (typically `~/.ssh/id_rsa.pub`)
into the `publicKey` field.
{{< /hint >}}
```
wrapping in a note box

File: content/providers/provider-terraform/migrate-hcl.md [hunk 18]
```diff
@@ -0,0 +1,527 @@
...
+The Crossplane AWS provider configuration handles authentication. You must
+create a Kubernetes secret file to authenticate with your AWS account.
+
+Crossplane supports AWS authentication with:
+Authentication Keys
+Web Identity
+Service Accounts
+
+For more information on cloud provider authentication, checkout the Provider
+Azure and Provider GCP authentication documentation.
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:441]:
```suggestion
{{< hint "note" >}}
For more information on cloud provider authentication, read the 
[Provider Azure]({{<ref "providers/provider-azure/authentication">}}) or [Provider GCP]({{<ref "providers/provider-gcp/authentication">}}) authentication documentation.
{{< /hint >}}
```

File: content/providers/provider-terraform/migrate-hcl.md [hunk 19]
```diff
@@ -0,0 +1,527 @@
...
+
+## Authenticate with your cloud provider
+
+The Crossplane AWS provider configuration handles authentication. You must
+create a Kubernetes secret file to authenticate with your AWS account.
+
+Crossplane supports AWS authentication with:
+Authentication Keys
+Web Identity
+Service Accounts
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:438]:
```suggestion
The provider supports AWS authentication with:
* [Authentication Keys]({{<ref "providers/provider-aws/authentication#aws-authentication-keysAuthentication">}})
* [Web Identity]({{<ref "providers/provider-aws/authentication#webidentity">}})
* [Service Accounts]({{<ref "providers/provider-aws/authentication#iam-roles-for-service-accounts">}})
```

File: content/providers/provider-terraform/migrate-hcl.md [hunk 20]
```diff
@@ -0,0 +1,527 @@
...
+
+Create a new Kubernetes secret.
+
+```shell
+kubectl -n upbound-system create secret generic aws-creds --from-file=credentials=aws-credentials
+```
+
+Verify your secret with `kubectl describe secret`.
+
+```shell
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:451]:
```suggestion
```shell {copy-lines="1"}
```

File: content/providers/provider-terraform/migrate-hcl.md [hunk 21]
```diff
@@ -0,0 +1,527 @@
...
+
+## Deploy your claim
+
+Next, apply your Crossplane Composition.
+
+```shell
+kubectl apply -f claim.yaml
+```
+
+## Verify your deployment
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:494]: You should add the `kubectl get claim...` output to show when it's ready instead of just waiting forever for SSH

File: content/providers/provider-terraform/migrate-hcl.md [hunk 22]
```diff
@@ -0,0 +1,527 @@
...
+ssh ubuntu@<> -i <path_to_key>
+```
+
+Update the IP address and key path with your instance public IP and the path to
+your local private key file (typically `~/.ssh/id_rsa`).
+
+## Recommended practices
+
+When you move from Terraform to Crossplane, the `provider-terraform` approach eases the transition. When you get more comfortable with Crossplane, the
+native Kubernetes spec configuration enables you to remove your original HCL
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:508]:
```suggestion
native cloud provider configuration enables you to remove your original HCL
```

File: content/providers/provider-terraform/migrate-hcl.md [hunk 23]
```diff
@@ -0,0 +1,527 @@
...
+resource definitions.
+
+Upbound's Marketplace can help you recreate your Terraform configurations with
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:511]:
```suggestion
The [Upbound Marketplace](https://marketplace.upbound.io) can help you recreate your Terraform configurations with
```

File: content/providers/provider-terraform/migrate-hcl.md [hunk 24]
```diff
@@ -0,0 +1,527 @@
...
+Crossplane providers. In the EC2 example, the Marketplace has a dedicated
+section for all the managed resources of the EC2 API group. The original
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:513]:
```suggestion
section for all the managed resources of the [EC2 API group](https://marketplace.upbound.io/providers/upbound/provider-aws-ec2/. The original
```

File: content/providers/provider-terraform/migrate-hcl.md [hunk 25]
```diff
@@ -0,0 +1,527 @@
...
+required by Terraform are the same values required by Crossplane.
+
+## Next steps
+
+You just created an EC2 instance with an SSH key with Crossplane. You created a
+Composite Resource Definition and a Composition from your original Terraform
+configuration.
+
+For more information on Upbound, Crossplane, and control planes, review the
+Crossplane Architecture Framework for best practices when building your infrastructure.
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:527]:
```suggestion
[Crossplane Architecture Framework]({{<ref "/xp-arch-framework">}}) for best practices when building your infrastructure.
```

**Review by plumbis — APPROVED** [review]
Suggested edits but no blockers

File: content/providers/provider-terraform/migrate-hcl.md [hunk 26]
```diff
@@ -0,0 +1,529 @@
...
+kind: Instance
+metadata:
+  name: example
+  namespace: default
+spec:
+  parameters:
+    region: us-east-1
+    keyPairName: test-keypair
+    publicKey: |
+      ssh-rsa AAAAB3NzaC1...
```
Comment (plumbis) [inline content/providers/provider-terraform/migrate-hcl.md:422]:
```suggestion
  parameters:
    region: us-east-1
    publicKey: |
      ssh-rsa AAAAB3NzaC1...
```

The change in the XRD doesn't require the name anymore.

**Review by plumbis — APPROVED** [review]
